openfecli/commands/gather.py

```diff
@@ -51,6 +84,114 @@ def legacy_get_type(res_fn):
         return 'complex'
 
 
+def _get_ddgs(legs):
+    import numpy as np
+    DDGs = []
+    for ligpair, vals in sorted(legs.items()):
+        DDGbind = None
+        DDGhyd = None
+        bind_unc = None
+        hyd_unc = None
+
+        if 'complex' in vals and 'solvent' in vals:
+            DG1_mag, DG1_unc = vals['complex']
+            DG2_mag, DG2_unc = vals['solvent']
+            if not ((DG1_mag is None) or (DG2_mag is None)):
+                # DDG(2,1)bind = DG(1->2)complex - DG(1->2)solvent
+                DDGbind = (DG1_mag - DG2_mag).m
+                bind_unc = np.sqrt(np.sum(np.square([DG1_unc.m, DG2_unc.m])))
+        elif 'solvent' in vals and 'vacuum' in vals:
+            DG1_mag, DG1_unc = vals['solvent']
+            DG2_mag, DG2_unc = vals['vacuum']
+            if not ((DG1_mag is None) or (DG2_mag is None)):
+                DDGhyd = (DG1_mag - DG2_mag).m
+                hyd_unc = np.sqrt(np.sum(np.square([DG1_unc.m, DG2_unc.m])))
+        else:  # -no-cov-
+            raise RuntimeError(f"Unknown DDG type for {vals}")
+
+        DDGs.append((*ligpair, DDGbind, bind_unc, DDGhyd, hyd_unc))
+
+    return DDGs
+
+def _write_ddg(legs, writer):
+    DDGs = _get_ddgs(legs)
+    writer.writerow(["ligand_i", "ligand_j", "DDG(i->j) (kcal/mol)",
+                      "uncertainty (kcal/mol)"])
+    for ligA, ligB, DDGbind, bind_unc, DDGhyd, hyd_unc in DDGs:
+        name = f"{ligB}, {ligA}"
+        if DDGbind is not None:
+            DDGbind, bind_unc = format_estimate_uncertainty(DDGbind,
+                                                            bind_unc)
+            writer.writerow([ligA, ligB, DDGbind, bind_unc])
+        if DDGhyd is not None:
+            DDGhyd, hyd_unc = format_estimate_uncertainty(DDGbind,
+                                                          bind_unc)
+            writer.writerow([ligA, ligB, DDGhyd, hyd_unc])
+
+def _write_raw_dg(legs, writer):
+    writer.writerow(["leg", "ligand_i", "ligand_j", "DG(i->j) (kcal/mol)",
+                     "uncertainty (kcal/mol)"])
+    for ligpair, vals in sorted(legs.items()):
+        name = ', '.join(ligpair)
+        for simtype, (m, u) in sorted(vals.items()):
+            if m is None:
+                m, u = 'NaN', 'NaN'
+            else:
+                m, u = format_estimate_uncertainty(m.m, u.m)
+            writer.writerow([simtype, *ligpair, m, u])
+
```

Comment:
```suggestion


```
Seriously I hate `black` but at this point we might have to actually make the jump.. at least with a 120 character limit (since OpenMM & RDKit have overly long method names).